docs/specification/draft/basic/authorization.mdx

```diff
@@ -302,6 +334,63 @@ Servers **MUST** return appropriate HTTP status codes for authorization errors:
+
+- **Minimum approach**: Include only the newly required scopes for the specific operation
+- **Recommended approach**: Include both existing relevant scopes and newly required scopes to prevent clients from losing previously granted permissions
+- **Extended approach**: Include existing scopes, newly required scopes, and related scopes that commonly work together
+
+The choice depends on the server's assessment of user experience impact and authorization friction. Servers **SHOULD** be consistent in their scope inclusion strategy to provide predictable behavior for clients.
+
+Servers **SHOULD** consider the user experience impact when determining which scopes to include in the response, as misconfigured scopes may require frequent user interaction.
+
+Example insufficient scope response:
+
+```http
+HTTP/1.1 403 Forbidden
+WWW-Authenticate: Bearer error="insufficient_scope",
+                         scope="files:read files:write user:profile",
+                         resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
+                         error_description="Additional file write permission required"
+```
+
+##### Client Scope Upgrade Flow
+
+Clients will receive scope-related errors during initial authorization or at runtime (`insufficient_scope`).
+Clients may respond to these errors by requesting a new access token with an increased set of scopes via an upgrade flow.
+Clients acting on behalf of a user **SHOULD** attempt the upgrade flow. Clients acting on their own behalf (`client_credentials` clients)
+**MAY** attempt the upgrade flow or abort the request immediately.
+
+The upgrade flow is as follows:
+
+1. **Parse error information** from the authorization server response or `WWW-Authenticate` header
+2. **Determine required scopes** using the following algorithm:
+   - If the `scope` parameter is present in the `WWW-Authenticate` header, use those scopes exactly as specified (trusting the server to include any necessary existing scopes along with newly required scopes)
+   - If `scope` is not available, use all scopes listed in `scopes_supported` from the Protected Resource Metadata document
+3. **Initiate (re-)authorization** with the determined scope set
+4. **Retry the original request** with the new authorization no more than a few times and treat this as a permanent authorization failure
+
+Clients **SHOULD** implement retry limits and **SHOULD** cache scope upgrade attempts to avoid
+repeated failures for the same resource and operation combination. The cache should track failed
+authorization attempts by resource and the scopes that were denied, preventing clients from
+repeatedly requesting the same permissions that were previously denied and avoiding unnecessary
+authorization server requests for known-failing scenarios.
```

Comment:
I was thinking about this cache today. I'm wondering what the granularity should be...

So we have:
* HTTP method
* MCP method (call/tool)
* Specific tool being called
* Specific payload of specific tool being called

Really any of these could be the reason for a new set of scopes... how could a client really keep track of this?